AbsaOSS/hyperdrive
# Feature/publish on maven 2

**Remarks**
- The setup doesn't currently work yet, because the project is not yet approved to be deployed on ossrh.
- To deploy with sources, javadoc and gpg signatures, run
`mvn clean deploy -Pdeploy`
- The modules _driver_, _ingestor-default_ and _shared_ are skipped and not deployed.
- Username and password for ossrh are expected to be provided at deployment.

**Test procedure with local Nexus**
**Part I: Deploy artifacts**
I tested deployment locally with Nexus Repository Manager OSS (This is an open source version)
1. Download and install Nexus Repository Manager OSS https://www.sonatype.com/download-oss-sonatype. 
Setup user with username _testusername_ and password _testpassword_

2. Create `~/.m2/settings.xml` file with following content
```
<settings>
        <servers>
                <server>
                        <id>ossrh</id>
                        <username>testusername</username>
                        <password>testpassword</password>
                </server>
        </servers>
</settings>
```

3. Use following distributionManagement in `<root>/pom.xml`
```
    <distributionManagement>
        <snapshotRepository>
            <id>ossrh</id>
            <url>http://localhost:8081/repository/maven-snapshots/</url>
        </snapshotRepository>
        <repository>
            <id>ossrh</id>
            <url>http://localhost:8081/repository/maven-releases</url>
        </repository>
    </distributionManagement>
```

4. Create Release version
`mvn versions:set -DnewVersion=1.0.0 -DgenerateBackupPoms=false`

5. Deploy
`mvn clean deploy -Pdeploy -DskipStaging=true`
Hint: `skipStaging` is **only required for local testing**, because Nexus OSS doesn’t support staging.

6. Check artifacts are published on local nexus
Check `http://localhost:8081/#browse/browse:maven-releases`

**Part II: Download and install Component-Archetype**

1. Use empty local repository
`mv ~/.m2/repository ~/.m2/repository_old`

2. Get dependency
`mvn dependency:get -Dartifact=za.co.absa.hyperdrive:component-archetype:1.0.0 -DremoteRepositories=ossrh::::http://localhost:8081/repository/maven-releases`
Hint: The `-DremoteRepositories`flag is **only required for local testing**.

3. Update local archetype catalog
`mvn archetype:crawl`

4. Generate the archetype
`cd /tmp/archetype`
`mvn -llr archetype:generate -DarchetypeGroupId=za.co.absa.hyperdrive -DarchetypeArtifactId=component-archetype -DarchetypeVersion=1.0.0 -DgroupId=com.myapp -DartifactId=mycomponent -Dversion=0.1-SNAPSHOT`

Hint: `-llr` alias `—legacy-local-repository` is **only required for local testing**.

5. Package generated project
`cd /tmp/archetype/mycomponent`
Add to `/tmp/archetype/mycomponent/pom.xml`
```
<repositories>
         <repository>
                <id>localnexus</id>
                <url>http://localhost:8081/repository/maven-releases</url>
         </repository>
 </repositories>
```
`mvn package`

6. Check generated project
Check `/tmp/archetype/mycomponent/target/mycomponent-0.1-SNAPSHOT.jar` exists
